tests/ga/extension_actor.py

```diff
@@ -0,0 +1,465 @@
+        Creates a ExtensionActor emulating an extension specified by the provided
+        extensionInfo and actionSet. Copies the dataFetcherBase to serve as the dynamic
+        loader of info required by a WireProtocolData* object.
+        """
+        self.action_scope = Mock()
+        ExtensionActor._configure_action_scope(self.action_scope, actionSet)
+
+        self.data_fetcher = dataFetcherBase.copy()
+        ExtensionActor._configure_data_fetcher(self.data_fetcher, actionSet, extensionInfo)
+
+        self.get_command = lambda cmd: getattr(self.action_scope, actionSet.getKeyForCommand(cmd), None)
+        self.extension_info = extensionInfo
+
+        self.statusBlobs = []
+        
+    def patch_popen(self):
+        """
+        Returns a mock that patches the existing subprocess.Popen function call (on start()) such that any
+        calls supposed to invoke the extension emulated by this object are intercepted and correctly ran
+        (via the action funcs in the actionSet that initialized this object). Any other commands are passed
+        through to the prior existing subprocess.Popen.
+
+        Note that the passthrough on failure to match a command to this extension's action commands enables
+        multiple patch_popen() Mocks to function at the same time.
+
+            try:
+                first_patch = first_ext.patch_popen()
+                first_patch.start()
+                second_patch = second_ext.patch_popen()
+                second_patch.start()
+
+                # Execute code that might try to invoke either extension.
+                {more code}
+
+            finally:
+                first_patch.stop()
+                second_patch.stop()
+
+        
+        For instance: if {more code} involves a invocation of the extension emulated by first_ext, the mock popen for
+        second_ext will be invoked, but will fail to match the provided command to one within second_ext's actionScope.
+        It will then fall back to the prior popen-- in this case, that's the mock popen from first_ext-- which will
+        successfully match the provided command to one within first_ext's actionScope. The corresponding action will
+        then be called and returned.
+
+        Note the need to interleave the patch_popen() and start() calls; in order for second_patch to know to fall back
+        to first_patch (instead of simply falling back to the original subprocess.Popen), first_patch must have already
+        replaced subprocess.Popen, which is handled by the start() function.
+        """
+
+        original_popen = subprocess.Popen
+
+        def mock_popen(command, *args, **kwargs):
+            # The command can be specified as a list of args or one single string. Handle that here.
+            format_command = lambda cmd: " ".join(cmd) if isinstance(cmd, list) else cmd
+
+            script_name = format_command(command).split("/")[-1] # The command passed here (if it corresponds to one within
```

Comment:
this assumes a specific format for the command. it may be working now in those tests, but if we happen to add another call to popen somewhere in the call stack this will break

Reply:
Good point. In fact, line 453 is a mitigation of one such situation (i.e. another call to popen in the call stack), wherein popen was passed a list instead of a string. However, I think there's only a finite set of possible args to popen (I think consisting of list and string).

Note, though, that a string (or list) that doesn't follow the specific format will simply fallback to the original popen (via the default on line 463). Is there another type(command) I need to account for?